src/ip-fragment.c

```diff
@@ -0,0 +1,181 @@
+/* Copyright 2016 Outscale SAS
+ *
+ * This file is part of Butterfly.
+ *
+ * Butterfly is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as published
+ * by the Free Software Foundation.
+ *
+ * Butterfly is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Butterfly.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <packetgraph/ip-fragment.h>
+#include <rte_config.h>
+#include <rte_ip.h>
+#include <rte_ip_frag.h>
+#include <rte_ether.h>
+#include "utils/mempool.h"
+#include "utils/bitmask.h"
+#include "packets.h"
+#include "brick-int.h"
+
+struct pg_ip_fragment_config {
+	enum pg_side output;
+	uint32_t mtu_size;
+};
+
+struct pg_ip_fragment_state {
+	struct pg_brick brick;
+	enum pg_side output;
+	struct rte_mbuf *pkts_out[64];
+	uint32_t mtu_size;
+};
+
+struct eth_ipv4_hdr {
+	struct ether_hdr eth;
+	struct ipv4_hdr ip;
+} __attribute__((__packed__));
+
+static struct pg_brick_config *ip_fragment_config_new(const char *name,
+						enum pg_side output,
+						uint32_t mtu_size)
+{
+	struct pg_brick_config *config = g_new0(struct pg_brick_config, 1);
+	struct pg_ip_fragment_config *ip_fragment_config =
+		g_new0(struct pg_ip_fragment_config, 1);
+
+	ip_fragment_config->output = output;
+	ip_fragment_config->mtu_size = mtu_size;
+	config->brick_config = (void *) ip_fragment_config;
+	return pg_brick_config_init(config, name, 1, 1, PG_DIPOLE);
+}
+
+static int do_fragmentation(struct pg_ip_fragment_state *state,
+			    struct rte_mbuf *pkt, struct pg_brick_edge *edge,
+			    enum pg_side from, struct pg_error **errp)
+{
+	uint64_t mask;
+	int32_t nb_frags;
+	struct  eth_ipv4_hdr *pkt_buf = rte_pktmbuf_mtod(pkt,
+							 struct eth_ipv4_hdr *);
+	struct ether_hdr eth = pkt_buf->eth;
+	struct rte_mbuf **pkts_out = state->pkts_out;
+	int l2_size = pkt->l2_len;
+	int ret;
+
+	rte_pktmbuf_adj(pkt, l2_size);
+	/* Because the ip rcf ask for the fragment size to be a multiple of 8,
+	* and rte_ipv4_fragment_packet don't check if the fragment size is
+	* a multiple of 8, we need to be sure to give a multiple of 8 for the
+	* 4rd argument*/
+	nb_frags = rte_ipv4_fragment_packet(pkt, pkts_out, 64,
+					    state->mtu_size - l2_size -
+					    (8 - ((sizeof(struct ipv4_hdr) +
+						   l2_size) % 8)),
+					    pg_get_mempool(), pg_get_mempool());
+	if (unlikely(nb_frags < 0)) {
+		*errp = pg_error_new_errno(-nb_frags,
+					   "fail to fragment packet");
```

Comment:
it's quite hard to fail a whole burst for a non-fragmentable packet, we can let the paquet pass, maybe it will be dropped by some other bricks after.